runtime/near-vm-logic/src/vmstate.rs

```diff
@@ -0,0 +1,177 @@
+        // This check is redundant in the sense that read_memory will perform it
+        // as well however it’s here to validate that `len` is a valid value.
+        // See documentation of MemoryLike::read_memory for more information.
+        self.0.fits_memory(offset, len).map_err(|_| HostError::MemoryAccessViolation)?;
+        let mut buf = vec![0; len as usize];
+        self.0.read_memory(offset, &mut buf).map_err(|_| HostError::MemoryAccessViolation)?;
+        Ok(buf)
+    }
+
+    /// Copies data from provided buffer into guest memory accounting for gas.
+    pub(super) fn set(
+        &mut self,
+        gas_counter: &mut GasCounter,
+        offset: u64,
+        buf: &[u8],
+    ) -> Result<()> {
+        gas_counter.pay_base(write_memory_base)?;
+        gas_counter.pay_per(write_memory_byte, buf.len() as _)?;
+        self.0.write_memory(offset, buf).map_err(|_| HostError::MemoryAccessViolation.into())
+    }
+
+    memory_get!(u128, get_u128);
+    memory_get!(u32, get_u32);
+    memory_get!(u16, get_u16);
+    memory_get!(u8, get_u8);
+    memory_set!(u128, set_u128);
+}
+
+/// Registers to use by the guest.
+///
+/// Provides interface to access registers while correctly accounting for gas
+/// usage.
+///
+/// See documentation of [`Memory`] for more motivation for this struct.
+#[derive(Default, Clone)]
+pub(super) struct Registers(std::collections::HashMap<u64, Box<[u8]>>);
+
+impl Registers {
+    /// Returns register with given index.
+    ///
+    /// Returns an error if (i) there’s not enough gas to perform the register
+    /// read or (ii) register with given index doesn’t exist.
+    pub(super) fn get<'s>(
+        &'s self,
+        gas_counter: &mut GasCounter,
+        register_id: u64,
+    ) -> Result<&'s [u8]> {
+        if let Some(data) = self.0.get(&register_id) {
+            gas_counter.pay_base(read_register_base)?;
+            gas_counter.pay_per(read_register_byte, data.len() as u64)?;
+            Ok(&data[..])
+        } else {
+            Err(HostError::InvalidRegisterId { register_id }.into())
+        }
+    }
+
+    /// Returns length of register with given index or None if no such register.
+    pub(super) fn get_len(&self, register_id: u64) -> Option<u64> {
+        self.0.get(&register_id).map(|data| data.len() as u64)
+    }
+
+    /// Sets register with given index.
+    ///
+    /// Returns an error if (i) there’s not enough gas to perform the register
+    /// read or (ii) if setting the register would violate configured limits.
+    pub(super) fn set(
+        &mut self,
+        gas_counter: &mut GasCounter,
+        config: &VMLimitConfig,
+        register_id: u64,
+        data: impl Into<Box<[u8]>>,
+    ) -> Result<()> {
+        let data = data.into();
+        let data_len = data.len() as u64;
+        gas_counter.pay_base(write_register_base)?;
+        gas_counter.pay_per(write_register_byte, data_len)?;
```

Comment:
This is somewhat problematic I suspect. Before this PR the allocation would have been accounted for by the memory read operation, but now that it does no longer allocate and its cost is likely to go down significantly, the cost of the allocation shifts to the register set operation instead. We should confirm that the register set operation is actually priced appropriately for what it now does.

cc @jakmeier

---

Ideally we would also convert after charging the gas for the operation. Something like this:

```suggestion
        gas_counter.pay_base(write_register_base)?;
        let len = u64::try_from(data.len()).map_err(|_| HostError::MemoryAccessViolation)?;
        gas_counter.pay_per(write_register_byte, len)?;
        let data = data.into();
```

But the fact that `data` is generic generic makes it somewhat non-straightforward. I fiddled with various combinations of libstd generics here (`Into<Box<[u8]>> + Into<&[u8]>`, `where T: Into<Box<[u8]>>, &T: IntoIterator, &T::IntoIter: ExactSizeIterator`) and I think the ideal approach here would be to add a custom trait basically. Something like this:

```rust
trait RegisterData: Into<Box<[u8]>> {
    fn len(&self) -> usize;
}
impl<'a> RegisterData for &'a [u8] { ... }
impl RegisterData for Vec<u8> { ... }
// impl whatever other types we use as a register data

// ---

    pub(super) fn set(
        &mut self,
        gas_counter: &mut GasCounter,
        config: &VMLimitConfig,
        register_id: u64,
        data: impl RegisterData,
    ) -> Result<()> {
```

Reply:
The way I’m thinking about gas here is that we should be assuming allocation always happens so estimator should test for read with allocation and similarly for write with allocation.  Allocations cannot always be avoided so we still need to be pessimistic about it.  Then again, the downside is that we’re leaving CPU time on the table if we are too pessimistic.

Reply:
Regarding traits, `Into<Box<[u8]>> + AsRef<[u8]>` should be enough.

Reply:
Thanks for the ping @nagisa ! Indeed we need to update the estimation, as continuous estimation showed -50% estimated gas cost for `WriteMemoryByte` this morning when comparing with the version from last Friday.

As @mina86 mentioned, the estimator should be updated to cover the new worst case. I filed https://github.com/near/nearcore/issues/8244, I don't think it's super urgent to fix but we need to track it.